### ISSUE TITLE
All Text Modules restricted to one or more Groups appears before choosing a Group

### ISSUE THREAD

**Issue by Kevin202972** [zammad/zammad#4890]
### Used Zammad Version

6.1.0-1696849338.59fe2e15.jammy

### Environment

- Installation method: Package
- Operating system (if you're unsure: `cat /etc/os-release` ): Ubuntu 22
- Database + version:  [e.g. postgresql 9.3, mysql 5.7, mariadb 10.3] postgresql 14
- Elasticsearch version: 7.17.13
- Browser + version: all Browsers

### Actual behaviour

If i create a ticket as an Agent i can use every text module. Every text module i have is restricted to one group. But i can use every text module when i select no group
![Screenshot 2023-10-18 062458](https://github.com/zammad/zammad/assets/142332276/03e6bb9c-2472-4ed1-841b-613f962820c3)

When i select a group, than i can use only the text moduls for this group
![Screenshot 2023-10-18 062524](https://github.com/zammad/zammad/assets/142332276/27f0112a-cf37-4caa-9e2c-99408d1b9b99)

### Expected behaviour

When no group is select, i should not see every text module

### Steps to reproduce the behaviour

1. Create two or more groups
2. Create text modules and restric them to the different groups
3. Create a ticket as an Agent and select no group 
4. use :: to use a text modules and now you can use every text module in the zammad instance 

### Support Ticket
Ticket#10169121

https://community.zammad.org/t/text-modules-restricted-to-one-group-appears-before-choosing-a-gorup/12367

### I'm sure this is a bug and no feature request or a general question.

yes

**Comment by rolfschmidt** [comment]
Not sure if this is a bug, actually. I tested it in stable-6.0 and there the behaviour was the same, so no regression here. We could change this issue to an enhancement to provide more options to define the behaviour for non-group selection.

I think this is how zammad behaves:

No group: Show all text modules
Selected group: Filter on group based text modules.

Maybe @martini can drop some note if he wants to change the behaviour idk.

**Comment by Kevin202972** [comment]
If it is no bug, than it would be very nice, if you can add an option, to hide all text modules, when no group is select or add the option to make it aviable for all by creating a text module. It would be good for our usecase, because it wouldn't be optimal if other departments saw that some of the answers we wrote were standard answers and vice versa

**Comment by cautopartes** [comment]
Our [documentation](https://admin-docs.zammad.org/en/latest/manage/text-modules.html) states:

```
If needed, you can restrict text modules to specific groups. With this, you can easily keep text module lists short and dedicate specific texts to where they belong.

You can adjust the group memberships for text modules at any time. This allows you to have the text module available globally (no groups selected) or one or several specific groups.
```

* It does not state that restricting text modules to groups has no effect in the ticket creation mask if no group is selected. 
* Additionally, as the admin explicitly configures a text module to be shown only for one, or several groups, this should be honored also in ticket creation.

**Configuration** (Admin > Manage > Text Modules)
Selecting no group makes text module available for all groups

**Ticket Creation** (**+** sign, bottom left in the UI)
Selecting no group makes text module available, with no group selected

* As ticket creation is usually not an admin task, I would expect the admins' configuration to take effect here, not the fact that the agent has not set a group yet "overwriting" this configuration. 
* Especially if the content of the text modules should not be disclosed to all groups, you can sneak through all systemwide available text modules by creating a ticket, not selecting a group and guessing name and keyword (by which they are found in the selection).

In summary, I personally would see it more like a tiny disclosure of present text modules and a loophole, than an enhancement.

**Comment by jh-due** [comment]
since it was considered an issue [here ](https://github.com/zammad/zammad/issues/4358)

I guess it is a bug

**Comment by cautopartes** [comment]
> since it was considered an issue [here ](https://github.com/zammad/zammad/issues/4358)
> 
> I guess it is a bug

Good point! Good research! I guess it is a bug, too.

**Comment by dominikklein** [comment]
@YetAnotherGerrit We need only from product perspective which way we are seeing in the future? Should text modules with a configured group not be part of the list of text modules when in the ticket create screen no group is selected yet (like described in this issue)?
When yes we need only to keep in mind to add this as a breaking change in the next release and not doing an backport of this change.

**Comment by mantas** [comment]
On the other hand, if text module has a group selected, some people may expect that text module to be visible to people who have access to that group only. Then it's pretty close to a data leak and could be backported.

**Comment by YetAnotherGerrit** [comment]
Most people fill out the ticket creation mask from top to bottom, so if we remove text modules that are group specific before group selection, we force the user into a non-linear workflow. This will be confusing.

But I see the use-case of the issuer, as having the text modules filtered might be beneficial for finding the correct text module in a really large list. And there will be situations, where not all agents should be able to see all text modules for privacy reasons.

My preferred long-term solution would be to finally allow ticket attributes to appear above title/text, which might include the group selection. But this is not the correct time and place for that.

So, for now, I suggest adding an admin option (maybe in the composer settings?) to toggle the behaviour:
- Group restricted text modules are all available while no group is selected (current behaviour, default)
- Group restricted text modules are only available if referred group is selected

**Comment by dominikklein** [comment]
Do we really want to go this way, we will then maybe have it "forever". And it's differently to other places where we have "group" restrictions.
The question is, which problem do we want to solve from business perspective? I think then we can check which is the best solution.

**Comment by mantas** [comment]
I'd rather tackle this as a data privacy issue with no setting. Setting for something like this looks like troublesome. And then we'd be forced to keep such iffy behaviour forver.

**Comment by YetAnotherGerrit** [comment]
As discussed in our refinement:
- While no group is selected, text modules that are restricted to groups only appear if the user has "create" permission to the according group
- Once a group is selected, text modules that are restricted to groups only appear if they match the selected group

**Comment by dominikklein** [comment]
As an addition we want to improve the general situation in the future, but for this we would need to change the form flow (e.g. re-ordering fields and move group before the body). Because this is a more advanced change, it's planned for a later step.

So we will fix the mentioned problem from @YetAnotherGerrit with this bug report and the other topic will be an enhancement for later.
